### PR TITLE
New Dashboard widgets for 2.10 ER1

### PR DESCRIPTION
**Pull request by josemigallas**
⚠️ This is not to be merged yet, we're using this branch as base.

**What this PR does / why we need it**:

We're removing the tabs in favor of two smaller widgets showing the top 5 last updated products and backends. Also removing the list of services from the Context Switcher and replacing it for a link to a Products and a Backend APIs index pages.

**Which issue(s) this PR fixes** 

[Address UI scalability issues in API Management Service](https://issues.redhat.com/browse/APPDUX-462)

**Verification steps** 

* Provider dashboard shows the 2 new widgets
* Products/Backend index pages can be visited:
  * Searchbar works
  * Pagination works

### CODE REVIEW

File: app/decorators/application_decorator.rb
Comment (Martouta) [inline app/decorators/application_decorator.rb:18]: We spoke about this 😄 Improvement in process ⚠️  👍

File: app/decorators/backend_api_decorator.rb
Comment (Martouta) [inline app/decorators/backend_api_decorator.rb:33]: I don't know how this is done internally, but if it does the following, it is scary as hell 😱
1. backend_api left outer join backend_api_config left outer join services
2. count distinct services.id where backend_api.id is this one

instead, we should do this:
1. backend_api left outer join backend_api_config
2. count distinct backend_api_config.service_id where backend_api.id is this one

File: app/decorators/service_decorator.rb
Comment (Martouta) [inline app/decorators/service_decorator.rb:76]: inverse situation to https://github.com/3scale/porta/pull/2194/files#r501110645

File: app/controllers/provider/admin/dashboards_controller.rb
Comment (Martouta) [inline app/controllers/provider/admin/dashboards_controller.rb:20]: if we use `paginate` instead as with the other controllers, how many would there be by default? is it worth it? in that case, it would be easier to DRY this up.

File: app/controllers/api/services_controller.rb
Comment (Martouta) [inline app/controllers/api/services_controller.rb:26]: we should definitely DRY this 😄

File: app/controllers/provider/admin/backend_apis_controller.rb
Comment (Martouta) [inline app/controllers/provider/admin/backend_apis_controller.rb:17]: let's DRY this 😄

File: app/controllers/provider/admin/dashboards_controller.rb
Comment (Martouta) [reply app/controllers/provider/admin/dashboards_controller.rb:20]: https://github.com/3scale/porta/pull/2194/files#r501113662

File: app/controllers/provider/admin/dashboards_controller.rb
Comment (Martouta) [inline app/controllers/provider/admin/dashboards_controller.rb:26]: https://github.com/3scale/porta/pull/2194/files#r501113662

File: app/controllers/provider/admin/backend_apis_controller.rb
Comment (Martouta) [inline app/controllers/provider/admin/backend_apis_controller.rb:17]: https://github.com/3scale/porta/pull/2194/files#r501113662

File: app/controllers/provider/admin/dashboards_controller.rb
Comment (Martouta) [inline app/controllers/provider/admin/dashboards_controller.rb:21]: https://github.com/3scale/porta/pull/2194/files#r501113662

File: app/decorators/service_decorator.rb
Comment (Martouta) [inline app/decorators/service_decorator.rb:72]: maybe a `sum` of all the `contracts_count` field in the DB for all the `application_plans` of this service.
that would be more efficient.

File: app/decorators/backend_api_decorator.rb
Comment (josemigallas) [reply app/decorators/backend_api_decorator.rb:33]: What about calling `service.backend_api_configs` like it's done in `Account > Backends`?
https://github.com/3scale/porta/blob/ed7ad0d57ab2c8de7d4241afd5008f89de31fc59/app/controllers/api/backend_usages_controller.rb#L16

File: app/decorators/service_decorator.rb
Comment (josemigallas) [reply app/decorators/service_decorator.rb:76]: Would `service.backend_api_configs` be valid here, like it's in `Account > Backends`?
https://github.com/3scale/porta/blob/ed7ad0d57ab2c8de7d4241afd5008f89de31fc59/app/controllers/api/backend_usages_controller.rb#L16

File: app/controllers/provider/admin/dashboards_controller.rb
Comment (josemigallas) [reply app/controllers/provider/admin/dashboards_controller.rb:20]: There won't be pagination in the dashboard, only the a list of 5 items.

File: app/controllers/api/services_controller.rb
Comment (josemigallas) [reply app/controllers/api/services_controller.rb:26]: > TLDR: similar code but no repetition.

Well, I think DRY is not always the best solution: 2 equal lines of code don't necessarily have to be repeated.
In this case and IMO the scopes are different enough (different models, different controllers) so that DRY is not applicable.
Also, today we order by the same column (updated_at), tomorrow we don't know, maybe hits.
As for `paginate`, I wouldn't obscure it behind some decorator but rather leave it explicit for a quicker reading (plus, we'd need to pass the params).
Finally, keep in mind that each to_json passes its own fields and methods.

File: app/decorators/backend_api_decorator.rb
Comment (guicassolato) [reply app/decorators/backend_api_decorator.rb:28]: I see what you're suggesting and indeed it would be an improvement:

```ruby
object = BackendApi.find(2)

object.services.count
# SELECT COUNT(*) FROM `services` INNER JOIN `backend_api_configs` ON `services`.`id` = `backend_api_configs`.`service_id` WHERE `backend_api_configs`.`backend_api_id` = 2

object.backend_api_configs.select(:service_id).distinct.count
# SELECT DISTINCT COUNT(DISTINCT `backend_api_configs`.`service_id`) FROM `backend_api_configs` WHERE `backend_api_configs`.`backend_api_id` = 2
```

However, I wonder if it's worth the effort. The proper way of optimising this would be by caching this count, but we don't have the time to do this now. Perhaps we can move on with this the way it is and consider another round of improvements later, with caching and all?

File: app/controllers/api/services_controller.rb
Comment (Martouta) [inline app/controllers/api/services_controller.rb:20]: If you are not going to use this variable outside this method, better make it a local variable of the method rather than an instance variable of the class that can be used all over the controller, the views and the helpers.

```suggestion
    raw_services = current_user.accessible_services
```

Comment (Martouta) [reply app/controllers/api/services_controller.rb:20]: I see that you use it for the counter... then better `@total_services_count = ...`
The less 'global variables' with less 'global info', the better.

File: app/controllers/api/services_controller.rb
Comment (Martouta) [inline app/controllers/api/services_controller.rb:18]: I see many tips (in the logs) to improve the performance.

image

File: app/javascript/src/BackendApis/components/IndexPage.js
Comment (Martouta) [inline app/javascript/src/BackendApis/components/IndexPage.js:134]:
```suggestion
                <TextInput placeholder="Find a Backend" name="findBackend" id="findProduct" type="search" aria-label="Find a backend" />
```

🙈

**Comment by thomasmaas** [comment]
Looking good! 

The only thing we didn't really think about is that the current product shows up as the title of the context switcher, which is a bit weird as individual products/backends are no longer part of it.
Screenshot 2020-10-14 at 14 20 15

File: app/controllers/api/services_controller.rb
Comment (Martouta) [reply app/controllers/api/services_controller.rb:19]: For a following PR 👍

File: app/controllers/api/services_controller.rb
Comment (Martouta) [reply app/controllers/api/services_controller.rb:22]: For a following PR 👍

File: app/decorators/application_decorator.rb
Comment (Martouta) [inline app/decorators/application_decorator.rb:17]: Where is all this called/used from? 🤔

File: app/decorators/backend_api_decorator.rb
Comment (Martouta) [reply app/decorators/backend_api_decorator.rb:28]: Yep, makes sense all of it! 💪  👍

File: app/decorators/service_decorator.rb
Comment (Martouta) [reply app/decorators/service_decorator.rb:67]: similarly to https://github.com/3scale/porta/pull/2194/files#r504635406, next round for improvements 👍

File: app/models/backend_api.rb
Comment (Martouta) [inline app/models/backend_api.rb:14]: As I mentioned in the description of https://github.com/3scale/porta/pull/2204, we'll later have to polish all this and add tests 👍

File: test/integration/api/services_controller_test.rb
Comment (Martouta) [inline test/integration/api/services_controller_test.rb:19]: later we will have to improve this test

File: test/integration/provider/admin/backend_apis_controller_test.rb
Comment (Martouta) [inline test/integration/provider/admin/backend_apis_controller_test.rb:19]: later we will have to improve this test

**Review by Martouta — DISMISSED** [review]
Great work! 👏 It can be polished but let's consider it for another round knowing that this feature is for today 😉

**Review by thomasmaas — DISMISSED** [review]
approved based on click testing in preview. search wasn't working yet due to indexing.

**Comment by cntlsn** [comment]
looking good, great work!
a couple of things I noticed while clicking around in preview:
* In the two widgets in the Dashboard, clicking on a Product or Backend name wouldn't lead you to the relative Product or Backend page
* In both the Product Dashboard widget and the Product page, when clicking on the 'Edit' item of the kebab menu the page reloads instead of leading you to the Edit Product page.

**Review by damianpm — CHANGES_REQUESTED** [review]
There are two issues to fix:
1.- Icons are not equal in Dashboard
icons
2.- There is an error in `isPlainButtonAction`, this may be the cause you can't click on API names:
Screenshot 2020-10-14 at 15 05 51

File: app/decorators/application_decorator.rb
Comment (josemigallas) [reply app/decorators/application_decorator.rb:17]: This parses `updated_at` to a human friendly format, depending on the user's time zone when calling `to_json`.

**Comment by josemigallas** [comment]
> There are two issues to fix:
> 1.- Icons are not equal in Dashboard
> icons
> 2.- There is an error in `isPlainButtonAction`, this may be the cause you can't click on API names:
> Screenshot 2020-10-14 at 15 05 51

1. Good catch, fixed 👍 
2. I added a comment there, it's supposed to be valid according to PF documentation.. but it might has been removed recently. It shouldn't be a problem, we can take a look at it afterwards.

**Comment by thomasmaas** [comment]
@josemigallas I'd rather fix the broken links before merging, so we can deploy master to saas if we have to.

File: app/javascript/src/Dashboard/components/APIDataListItem.jsx
Comment (damianpm) [inline app/javascript/src/Dashboard/components/APIDataListItem.jsx:68]: I think we can simply remove this prop. According to PF docs is not required and I don't see any difference in the rendered output
pf-docs
```suggestion
```

Comment (josemigallas) [reply app/javascript/src/Dashboard/components/APIDataListItem.jsx:68]: OK sounds good to me!

**Review by damianpm — DISMISSED** [review]
LGTM, please run tests when ready

**Review by Martouta — DISMISSED** [review]

**Review by thomasmaas — APPROVED** [review]
All mentioned issues looking good now in preview

**Review by Martouta — APPROVED** [review]

File: features/step_definitions/applications/applications_steps.rb
Comment (Martouta) [inline features/step_definitions/applications/applications_steps.rb:171]:
```suggestion
  click_on @application.service.name
```

Looks more dynamic 🤔 

No need to be today 👍

File: features/old/stats/provider_side.feature
Comment (Martouta) [inline features/old/stats/provider_side.feature:18]: Ugly but let's polish later

**Review by Martouta — APPROVED** [review]
Great job! 👏

**Review by Martouta — APPROVED** [review]

**Review by Martouta — DISMISSED** [review]

**Review by Martouta — APPROVED** [review]

File: features/old/services/switch.feature
Comment (Martouta) [inline features/old/services/switch.feature:15]: This should be the `@wip` instead of the next one 🙈 My bad

**Review by Martouta — APPROVED** [review]

**Review by damianpm — APPROVED** [review]
🥳

**Pull request merged**
